### ISSUE TITLE
meta: Stabilize WASI support

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/tokio#4827]
This tracks outstanding issues to stabilize WASI support for Tokio.

Initial PR: #4716

Known issues:

* [ ] tests were disabled and need to be re-enabled
* [x] `tokio::time` can be stabilized.
* [x] Use `target_family = "wasm"` once MSRV becomes 1.54. Current workaround: [#4865](https://github.com/tokio-rs/tokio/pull/4865)
* [ ] WASI lacks many APIs needed for getting Tokio examples working
  * [ ] `TcpStream::connect`
  * [ ] `TcpListener::bind`

Refs: #1597

**Comment by rjzak** [comment]
Since #1597 was closed, perhaps track the differences between the three WebAssembly targets supported by Rust:
* wasm32-unknown-emscripten
* wasm32-unknown-unknown
* wasm32-wasi

Also, add an integration test which:
* Compiles a test application in Rust to Wasi which accepts connections
* Runs the test Wasi application (via Wasmtime library)
* Connect and ensure it worked
* Cleanly shutdown the Wasi app

**Comment by MediosZ** [comment]
For missing APIs like TcpStream and TcpListener, I tried to replace std with [wasmedge_wasi_socket](https://github.com/second-state/wasmedge_wasi_socket) and it works.
But this requires using WasmEdge and is not compatiable with other runtimes.
Some refs:
[mio](https://github.com/MediosZ/mio/commit/b73a96c709cc235c49a4628fe7f6929d428141e3)
[tokio](https://github.com/MediosZ/tokio/commit/ba1555ac6214d8e481a002f8229ac6f426ecc195)

**Comment by umgefahren** [comment]
Since WASM Threads are already support by Rust Std (although in nightly with build std) would it be possible to bring the multi thread runtime to WASM?

**Comment by Darksonn** [comment]
I'm sure we will support it eventually, but it is not a priority right now.

**Comment by frewsxcv** [comment]
> Since WASM Threads are already support by Rust Std (although in nightly with build std) would it be possible to bring the multi thread runtime to WASM?

@umgefahren Do you have a link to the PR where this was implemented in std?

**Comment by rjzak** [comment]
There is some code for Wasm threads, but most of it is empty or unsupported: [wasm/atomics/thread.rs](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/wasm/atomics/thread.rs).

The same is true for Wasi, [wasi/thread.rs](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/wasi/thread.rs).

There has been some progress on Wasi threads, the Wasi-Libc project has some pthread functions https://github.com/WebAssembly/wasi-libc/search?q=pthread, but most seem to be in the musl library. Even if Rust or C had code for Wasi threads, this still requires a Wasi runtime to be able to use it, and I don't think we're there yet.

**Comment by milesrichardson** [comment]
has anyone tried somehow replacing threads with web workers? intuitively that seems like it would work, but I don't know enough about Tokio or Rust to say for certain

**Comment by jeffparsons** [comment]
FYI: Wasmtime just recently landed thread support: https://github.com/bytecodealliance/wasmtime/pull/5484 (not yet in any release)

**Comment by BarbossHack** [comment]
"Announcing wasi-threads" (proposal)
https://bytecodealliance.org/articles/wasi-threads

**Comment by NobodyXu** [comment]
I wonder if [WASIX](https://wasmer.io/posts/announcing-wasix) would also be worth supporting?

It looks like they also have their own forked version of tokio for this https://github.com/wasix-org/tokio/tree/wasix-1.24.2

**Comment by BarbossHack** [comment]
« **WASI threads, implementation of wasm32-wasi-preview1-threads target**

This PR adds a target proposed in rust-lang/compiler-team#574 by @abrown and implementation of std::thread::spawn for the target wasm32-wasi-preview1-threads »

https://github.com/rust-lang/rust/pull/112922

**Comment by jeffparsons** [comment]
> I wonder if [WASIX](https://wasmer.io/posts/announcing-wasix) would also be worth supporting?

See also the open issue about supporting WasmEdge, which seems to be a similar situation: both have their own Tokio fork, but presumably would like to have support upstream in Tokio so they don't need to tell their users to put in an override.

I anticipate that some of the hunger for runtime-vendor-specific extensions _from end users of Tokio_ will fade as WASI Preview 2 comes together later this year, but I definitely don't understand the space well enough to say for sure, or if the vendors themselves would still hope for some upstream support. E.g. if a vendor adds a special IO primitive to their runtime that is faster than what's in WASI by accepting a more relaxed security posture, they might want their users to be able to opt in to using that.

So I guess the question on my mind is one of policy: would the Tokio project want to carry vendor-specific patches? For small things only? For "support WASI Preview 1 long term even after Preview 2 ships"? Yes, but only until WASI itself hits 1.0?

**Comment by NobodyXu** [comment]
> WASI Preview 2 comes together later this year

That will be great, hopefully it will end this vendor-specific extension.
Or even better, perhaps wasm should consider supporting vendor-specific extension officially.

**Comment by Darksonn** [comment]
We already have unstable support for wasi, and I don't mind PRs with updates to our unstable support as long as they aren't too complicated.

**Comment by jeffparsons** [comment]
> That will be great, hopefully it will end this vendor-specific extension.
> And perhaps wasm should consider supporting vendor-specific extension officially.

I don't think this is going to go away entirely, and it's not clear to me what additional support in Wasm for vendor extensions would look like beyond what is already afforded by the WebAssembly Component Model.

I suspect there will always be demand for compilers to support vendor-specific extensions, or _something equivalent_. E.g. Wasmer's WASIX provides POSIX-compatible APIs so that you can compile existing code bases (without invasive modifications) to Wasm, and then run them anywhere. To achieve this today, they have their own `rustc` fork, and their own `tokio` fork. And I suspect we're going to see more of that sort of thing from other vendors.

I wonder if this situation could be simplified by the Rust project itself owning a WIT [world](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#wit-worlds) that defines what `rustc` needs of a complete OS? Then `rustc` itself, `tokio`, etc., could target this world, and then WASIX and other compatibility layers can provide an adapter from this Rust-specific World to their API. The indirection can be compiled away, so I don't think the overhead of an extra layer would be an issue. Then instead of `rustc` and _L_ different libraries needing to carry support for _V_ different vendor-specific extensions, those extensions would simply implement one world defined by Rust.

Unless/until we ended up with a single standard "full OS" world (presumably) defined under the umbrella of the WebAssembly project you might still have a world defined by each language toolchain that's interested in this sort of thing (e.g. Zig, Nim, etc.) but an adapter component is a lot lighter-weight than managing an entire fork of `rustc` or Tokio! And it might nudge us in the direction of ending up with a single standard "full OS" WebAssembly world, just like how WASIX is nudging things in that direction. Then instead of _C_ compilers plus _L_ libraries all mapping to _V_ different vendors, you'd eventually be able to have each compiler or library map to _one_ world, and each vendor also mapping to that _one_ world.

Slight tangent: POSIX / "full OS" compatibility might be an anti-goal or at least orthogonal to the goals of a lot of people working in the WebAssembly space (the ambition goes a _long_ way beyond "cross platform bytecode for POSIX"), but there is clearly demand, so it might be best to embrace that. I could even imagine it being a "gateway drug" towards WASI adoption.

**Comment by banocean** [comment]
> WASI lacks many APIs needed for getting Tokio examples working
Technically APIs are there but most of them aren't yet part of WASI std and are still proposals. There is `wasi_ext` nightly feature too`.

**Comment by banocean** [comment]
I think there are few things that should be changed in tracking issue and taken into consideration
 - There is ongoing work on `wasm32-wasi-preview2` target
   - https://github.com/rust-lang/rust/pull/119616
   - https://github.com/rust-lang/compiler-team/issues/694)
 - WASI Preview 2 introduced [a lot of missing APIs as part of WASI specification](https://github.com/WebAssembly/WASI/blob/main/preview2/README.md#wasi-preview-2-contents)
 - There might still be APIs that still could be exposed but are blocked by "missing APIs" from the same feature (e.g. on preview 1 you couldn't create a new socket, but you could listen on one created earlier by runtime)
 - What about preview 3 plans for `stream` and `feature`

**Comment by Darksonn** [comment]
It's nice that wasi is working towards more socket options, but you say that they are still proposals, so it doesn't sound like there is anything for us to do yet?

**Comment by rylev** [comment]
I just wanted to drop in as one of the folks working on Wasi Preview 2 through the Bytecode Alliance. Hopefully I can clarify what the state of Wasi is, and what we, in the Bytecode Alliance, hope to accomplish in the medium term. 

Preview 2 (perhaps despite the name) is not merely a proposal, but is the official next evolution of Wasi (what has been known until now as Wasi has been retroactively renamed to Preview 1) and [has been officially adopted by the Wasi Subgroup](https://github.com/WebAssembly/WASI/tree/main/preview2#readme) which is an official subgroup of the WebAssembly Community Group of the W3C. Preview 2 is meant to hint towards the fact that the evolution will continue towards a Preview 3 before ultimately a 1.0. Even though the Wasi subgroup has chosen to not give this version of Wasi the 1.0 version, Preview 2 is meant as an "island of stability" around which the rest of ecosystem can grow. This means there is a commitment to support Preview 2 well beyond the introduction of Preview 3 and any other versions (on the order of magnitude of years) much like Preview 1 support has not suddenly ended with the introduction of Preview 2. 

Beyond this, there is significant interest in the Bytecode Alliance to see adoption of Preview 2 within the community which can help inform changes that will need to happen in future evolutions of Wasi. Members of the Bytecode Alliance have already begun implementing Preview 2 support in Tokio and dependent crates such as mio, and we hope that we'll be able to open pull requests soon and work with the Tokio maintainers on landing these changes. Preview 2 should have the required API support to implement everything Tokio needs for single threaded executors.

It is true that Preview 3 will likely have large impacts on the implementation details on Tokio support for Wasi (due to the introduction of first class async in the Component Model), but of course, there will be significant resources to help move Tokio over to Preview 3 when that times comes (which is probably earliest 1 year from now but may be longer).

**Comment by dicej** [comment]
FYI, I have created temporary forks of both `mio` and `tokio` (and also `tokio-postgres`, for real-world testing), adding WASI Preview 2 support -- most notably `wasi-sockets`, which is vastly more capable than Preview 1's meager socket support.  See https://github.com/dicej/wasi-sockets-tests for examples and links to the forks.

**Comment by dicej** [comment]
> It's nice that wasi is working towards more socket options, but you say that they are still proposals, so it doesn't sound like there is anything for us to do yet?

They're not just proposals.  The [`wasi-libc` implementation](https://github.com/WebAssembly/wasi-libc/pull/449) is feature-complete and in the process of being upstreamed incrementally.  I've created a temporary fork of Rust and added complete `wasi-sockets` support there, as well as the forks of `mio`, `tokio`, and `tokio-postgres` I mentioned above.  Meanwhile, @rylev has been [working to add `wasm32-wasi-preview2` support] to Rust, at which point we'll upstream the `wasi-sockets` changes to all those repos.  In short, the work is more or less done -- it's just logistics at this point.

One more fun tidbit: we've updated CPython as well, and got all the relevant `asyncio` tests passing (i.e. all except a few that spawn threads, fork processes, and/or use signals -- features not yet supported in WASI).

**Comment by banocean** [comment]
> Preview 2 (perhaps despite the name) is not merely a proposal, but is the official next evolution of Wasi

Propasal I was talking about is `wasm32-wasi-preview2` compile target 😅, the only thing to implement on tokio side should be changing old wasm/wasi related exclusions

Also all Preview 3 stuff was mentioned as a separate point becouse it's introducing stuff related to handling async

**Comment by Darksonn** [comment]
Thank you for clarifying. I have opened  #6323 to track support for preview 2.

**Comment by banocean** [comment]
> Thank you for clarifying. I have opened #6323 to track support for preview 2.

Can you elaborate what part of WASI support that is not part of WASI Preview 2 support

**Comment by Darksonn** [comment]
@banocean There are two aspects of this:

* Support WASI without requiring `--cfg tokio_unstable`.
* Support WASI preview 2. (We currently only support preview 1.)

This issue tracks the first aspect, the other issue tracks the second aspect.

**Comment by banocean** [comment]
> @banocean There are two aspects of this:
> 
>     * Support WASI without requiring `--cfg tokio_unstable`.
> 
>     * Support WASI preview 2. (We currently only support preview 1.)
> 
> 
> This issue tracks the first aspect, the other issue tracks the second aspect.

Now I'm kinda confused, is there any outline when a feature is put under `tokio_unstable` feature flag?

**Comment by Darksonn** [comment]
I don't know what you mean by "outline". Putting things under `--cfg tokio_unstable` is easy, since they are easily removed if they turn out to be problematic or incorrectly designed. Stabilizing them requires more serious thought.

**Comment by banocean** [comment]
> I don't know what you mean by "outline". Putting things under `--cfg tokio_unstable` is easy, since they are easily removed if they turn out to be problematic or incorrectly designed. Stabilizing them requires more serious thought.

Oh, I through that wasi support under tokio_unstable feature flag was base on the fact that introducing it normally won't comply with the 6 moths MSRV requirement

**Comment by yonas** [comment]
[Some projects](https://github.com/static-web-server/static-web-server/pull/205) have this as a blocker for supporting WASM.

What are the remaining issues before merge?

**Comment by banocean** [comment]
> [Some projects](https://github.com/static-web-server/static-web-server/pull/205) have this as a blocker for supporting WASM.
> 
> What are the remaining issues before merge?

It's not gonna break stuff since wasix has forked tokio and all wasix implementations are based on using that crate instead of the official tokio one

**Comment by Darksonn** [comment]
Supporting wasm in Tokio would require that someone who cares about wasm gets involved with Tokio and does the work to make it happen. To my knowledge, there are no blockers to [supporting wasi preview 2](https://github.com/tokio-rs/tokio/issues/6323) other than someone taking the time to implement whatever is missing. But I'm not going to do it — the wasm world is not something I have any involvement in, and we need someone who does have that knowledge to do it. 

(Note that the above talks about things like wasi that look like the platforms we support today. Non-web-worker browser wasm is a different story, and is out of scope for the foreseeable future.)

**Comment by dicej** [comment]
Now that https://github.com/rust-lang/rust/pull/129638 has been merged, I can start working on `mio` and `tokio` PRs to support `wasm32-wasip2`.  I'll try to get a draft PR up for `mio` this week, at least.

**Comment by banocean** [comment]
> Supporting wasm in Tokio would require that someone who cares about wasm gets involved with Tokio and does the work to make it happen. To my knowledge, there are no blockers to [supporting wasi preview 2](https://github.com/tokio-rs/tokio/issues/6323) other than someone taking the time to implement whatever is missing. But I'm not going to do it — the wasm world is not something I have any involvement in, and we need someone who does have that knowledge to do it.
> 
> (Note that the above talks about things like wasi that look like the platforms we support today. Non-web-worker browser wasm is a different story, and is out of scope for the foreseeable future.)

The blocker was rust waiting for wasi-libc preview 2 support, looks like it started moving. Only things that could stand in a way on tokio side is previous wasm compatibility or some system apis not getting covered.

**Comment by Darksonn** [comment]
Ah, I see. I did not know that it was blocked on adding support for things in std.

**Comment by dicej** [comment]
https://github.com/tokio-rs/mio/pull/1836
https://github.com/tokio-rs/tokio/pull/6893

**Comment by fbjork** [comment]
What is the latest progress on this issue? We are blocked on this.

**Comment by Darksonn** [comment]
It's the mio PR linked above.

**Comment by Darksonn** [comment]
The latest status is that preview 2 is supported thanks to the work by @dicej. Thanks!

**Comment by banocean** [comment]
> The latest status is that preview 2 is supported thanks to the work by [@dicej](https://github.com/dicej). Thanks!

Now we got to preview 3 release (https://bytecodealliance.org/articles/WASI-0.3) and wasm32-wasip3 getting promoted to tier 2 target (https://github.com/rust-lang/rust/pull/161940). Should there be another issue to keep track of support for preview 3 same way as there was one for preview 2 (#6323)?

**Comment by dicej** [comment]
Supporting p3 (at least to the same degree p2 is currently supported) in both `tokio` and `mio` should be easy, since `wasi-libc` takes care of the details of making them both look like POSIX-style platforms.  It _might_ not require any changes at all given that the current `#[cfg(not(all(target_os = "wasi", target_env = "p1")))]` guards will work just fine for p3.  I'll give it a try when I have a chance to see what's actually needed.

Beyond the p2 feature set, the plan is to add cooperative multithreading support in WASI 0.3.3 (WASI-SDK 34 already has experimental [pthread support](https://github.com/WebAssembly/wasi-sdk/blob/main/CoopThreading.md) which uses it), which will allow us to enable (non-preemptive) threading for the Rust `wasm32-wasip3` target and use it to support e.g. the `tokio` multithreaded runtime and the various features which rely on `spawn_blocking`, e.g. `tokio::fs`, `tokio::net::lookup_host`, etc.  Ideally, `tokio` wouldn't use `spawn_blocking` for such things on WASIp3, since it has native support for asynchronous filesystem operations, host lookup, etc., but it's a fine place to start since it requires minimal code changes.

**Comment by banocean** [comment]
> which will allow us to enable (non-preemptive) threading for the Rust wasm32-wasip3 target and use it to support e.g. the tokio multithreaded

So the plan is to wait til it becomes part of rust standard library for wasm32-wasip3 target?

>  Ideally, tokio wouldn't use spawn_blocking for such things on WASIp3, since it has native support for asynchronous filesystem operations, host lookup, etc., but it's a fine place to start since it requires minimal code changes

But at the same time doesn't it make us wait for something that at the end won't be needed and those changes will be made anyway?


Also a little note, thank you so much your replies they are really specific and helpful, it gives me so much better idea on where rust ecosystem is in terms of making wasi targets actually usable (thank you for work that you do to make it go forward too obviously)

**Comment by dicej** [comment]
> So the plan is to wait til it becomes part of rust standard library for wasm32-wasip3 target?

I'll try using `nightly` (which already has the `wasm32-wasip3` target) when I get a chance (i.e. before the Rust `wasm32-wasip3` target is stabilized), and post PRs for `mio` and `tokio` if changes are needed.  I imagine those changes will need to be feature-gated until `wasm32-wasip3` is available in the MSRV for both projects, but I guess it depends on what those changes end up being, exactly.

> But at the same time doesn't it make us wait for something that at the end won't be needed and those changes will be made anyway?

Given that Rust 1.100 (the first stable version with `wasm32-wasip3` support) will be released November 12th, and WASI 0.3.3 should be released in early December, I don't think it's a particularly long wait, and if you're willing to use `nightly` and pass some extra link flags to use the WASI-SDK cooperative multithreading sysroot, you might not have to wait at all.

I do think it would be fantastic for `tokio` to use native asynchronous filesystem etc. operations instead of `spawn_blocking` for `wasm32-wasip3`, but that will be a _big_ patch with a bunch of WASI-specific code paths, adding a significant maintenance burden.  If someone wants to tackle that in the short term (and is willing to take responsibility for ongoing maintenance), great, go for it.  I'm not that person, though -- at least not right now.

**Comment by dicej** [comment]
Alex just pointed me to [this](https://doc.rust-lang.org/nightly/rustc/platform-support/wasm32-wasip3.html#testing-cooperative-multithreading), which explains how to use `wasm32-wasip3` multithreading with `nightly`.  That works _today_, so in theory multithreaded `tokio` and `spawn_blocking` should be within reach for those willing to ride the bleeding edge.
